README.md

Please note that currently the documentation lags behind the development of the program. It will be 
improved over time. 

# Hilux Masternode Tool (DMT)

## Contents

 * [Masternodes](#masternodes)
 * [Hilux Masternode Tool](#hilux-masternode-tool)
   * [Feature list](#feature-list)
   * [Supported hardware wallets](#supported-hardware-wallets)
 * [Configuration](#configuration)
   * [Setting up the hardware wallet type](#setting-up-the-hardware-wallet-type)
   * [Connection setup](#connection-setup)
     * [Connection to a local node](doc/config-connection-direct.md)
     * [Connection to a remote node through an SSH tunnel](doc/config-connection-ssh.md)
     * [Connection to "public" JSON-RPC nodes](doc/config-connection-proxy.md)
   * [Masternode setup](#masternode-setup)
     * [Scenario A: moving masternode management from Hilux Core](doc/config-masternodes-a.md)
     * [Scenario B: configuring a new masternode](doc/config-masternodes-b.md)
   * [Command line parameters](#command-line-parameters)
 * [Features](#features)
   * [Starting a masternode](#starting-a-masternode)
   * [Registration of a deterministic masternode](doc/deterministic-mn-migration.md)
   * [Transferring masternode earnings](#transferring-masternode-earnings)
   * [Signing messages with a hardware wallet](#signing-messages-with-a-hardware-wallet)
   * [Changing a hardware wallet PIN/passphrase](#changing-a-hardware-wallet-pinpassphrase)
   * [Browsing and voting on proposals](doc/proposals.md)
   * [Hardware wallet initialization/recovery](doc/hw-initialization-recovery.md)
     * [Updating hardware wallet firmware](doc/hw-initr-update-firmware.md)
 * Building the DMT executables
    * [macOS](doc/build-dmt-mac.md)
    * [Windows](doc/build-dmt-windows.md)
    * Linux
       * [Ubuntu](doc/build-dmt-linux-ubuntu.md)
       * [Fedora](doc/build-dmt-linux-fedora.md)
 * [Downloads](https://github.com/Bertrand256/hilux-masternode-tool/releases/latest)
 * [Changelog](changelog.md)

## Masternodes

Hilux masternodes are full nodes which are incentivized by receiving a share of the block reward as payment in return 
for the tasks they perform for the network, of which the most important include participation in *InstantSend* 
and *PrivateSend* transactions. In order to run a masternode, apart from setting up a server which runs the 
software, you must dedicate 1000 Hilux as *collateral*, which is *"tied up"* in your node as long as you want 
it to be considered a masternode by the network. It is worth mentioning that the private key controlling the 
funds can (and for security reasons, should) be kept separately from the masternode server itself.

A server with the Hilux daemon software installed will operate as a Hilux full node, but before the rest of the 
network accepts it as a legitimate masternode, one more thing must happen: the person controlling the node 
must prove that they are also in control of the private key to the node's 1000 Hilux *collateral*. 
This is achieved by sending a special transaction to the network (ProRegTx), signed by this private key.

This action can be carried out using the *Hilux Core* reference software client. As can be expected, this 
requires sending 1000 Hilux to an address controlled by the *Hilux Core* wallet. After the recent increase 
in the value of Hilux and a burst in the amount of malware distributed over the Internet, you do not have 
to be paranoid to conclude that keeping large amounts of funds in a software wallet is not the most secure option. 
For these reasons, it is highly recommended using a **hardware wallet** for this purpose.

# Hilux Masternode Tool

The main purpose of the application is to give masternode operators (MNOs) the ability to 
send the ProRegTx with an easy to use a graphical user interface if the 
masternode collateral is controlled by a hardware wallet such as Trezor, KeepKey or Ledger.

## Feature list

* Sending the `ProRegTx` transaction if the collateral is controlled by a hardware wallet
* Transferring masternode earnings safely, without touching the 1000 Hilux funding transaction
* Signing messages with a hardware wallet
* Voting on proposals
* Initialization/recovery of hardware wallets seeds
* Updating of hardware wallets firmware (Trezor/KeepKey)
* Support for Hilux Testnet

## Supported hardware wallets

- [x] Trezor (model One and T)
- [x] KeepKey
- [x] Ledger Nano S

Most of the application features are accessible from the main program window:  
![Main window](doc/img/dmt-main-window-mn-list.png)

# Configuration

## Setting up the hardware wallet type
 * Click the `Configure` button.
 * Select the `Miscellaneous` tab in the configuration dialog that appears.
 * Depending on the type of your hardware wallet, select the `Trezor`, `KeepKey` or `Ledger Nano S` option.  
     ![Configuration window](doc/img/dmt-config-dlg-misc.png)

## Connection setup

Most of the application features involve exchanging data between the application itself and the Hilux network. To do this, *DMT* needs to connect to one of the full nodes on the network, specifically one which can handle JSON-RPC requests. This node plays the role of a gateway for *DMT* to the Hilux network. It does not matter which full node provides the service, because all nodes reach consensus by synchronizing information between each other on the Hilux network.

Depending on your preferences (and skills) you can choose one of three possible connection types:
 * [Direct connection to a local node](doc/config-connection-direct.md), for example to *Hilux Core* running on your normal computer.
 * [Connection to a remote node through an SSH tunnel](doc/config-connection-ssh.md), if you want to work with a remote Hilux daemon (like your masternode) through an SSH tunnel.
 * [Connection to "public" JSON-RPC nodes](doc/config-connection-proxy.md), if you want to use nodes provided by other users.

## Masternode setup

Here we make the following assumptions:
  * You already have a server running the Hilux daemon software (*hiluxd*) that you want to use as a masternode. If you don't, you will need to install and configure one first by following the guide on the [Hilux Wiki](https://docs.hilux.org/en/stable/masternodes/setup.html).
  * We occasionally refer to the *hiluxd* configuration file, so it is assumed that *hiluxd* is running under a Linux operating system (OS), which is the most popular and recommended OS for this purpose.
  * Your server has a public IP address that will be visible on the Internet.
  * You have set up a TCP port on which your *hiluxd* listens for incoming connections (usually 9999).

Further configuration steps depend on whether you already have a masternode controlled by *Hilux Core* which you want to migrate to a hardware wallet managed by *DMT*, or if you are setting up a new masternode.

[Scenario A - moving masternode management from Hilux Core](doc/config-masternodes-a.md)  
[Scenario B - configuration of a new masternode](doc/config-masternodes-b.md)

## Command line parameters

The application currently supports the following command-line parameters:
* `--data-dir`: a path to a directory in which the application will create all the needed files, such as: configuration file, cache and log files; it can be useful for users who want to avoid leaving any of the application files on the computer - which by default are created in the user's home directory - and instead to keep them on an external drive
* `--config`: a non-standard path to a configuration file. Example:
  `HiluxMasternodeTool.exe --config=C:\dmt-configs\config1.ini`



# Features

## Starting a masternode

Once you set up the Hilux daemon and perform the required *DMT* configuration, you need to broadcast the `start masternode` message to the Hilux network, so the other Hilux nodes recognize your daemon as a masternode and add it to the payment queue.

To do this, click the `Start Masternode using Hardware wallet` button.

### Sequence of actions

This section describes the steps taken by the application while starting the masternode, and possible errors that may occur during the process.

The steps are as follows:

1. Verification that all the required fields are filled with correct values. These fields are: `IP`, `port`, `MN private key`, `Collateral`, `Collateral TX hash` and `TX index`.
  An example message in case of errors:  
  ![Invalid collateral transaction id](doc/img/startmn-fields-validation-error.png)

2. Opening a connection to the Hilux network and verifying if the Hilux daemon to which it is connected is not still waiting for synchronization to complete.
  Message in case of failure:  
  ![Hilux daemon synchronizing](doc/img/startmn-synchronize-warning.png)

3. Verification that the masternode status is not already `ENABLED` or `PRE_ENABLED`. If it is, the following warning appears:  
  ![Warning: masternode state is enabled](doc/img/startmn-state-warning.png)  
  If your masternode is running, and you decide to send a `start masternode` message anyway, your masternode payment queue position will be reset.

4. Opening a connection to the hardware wallet. Message in case of failure:  
  ![Cannot find Trezor device](doc/img/startmn-hw-error.png)

5. If the `BIP32 path` value is empty, *DMT* uses the *collateral address* to read the BIP32 path from the hardware wallet.

6. Retrieving the Hilux address from the hardware wallet for the `BIP32 path` specified in the configuration. If it differs from the collateral address provided in the configuration, the following warning appears:  
  ![Hilux address mismatch](doc/img/startmn-addr-mismatch-warning.png)  
  The most common reason for this error is mistyping the hardware wallet passphrase. Remember that different passphrases result in different Hilux addresses for the same BIP32 path.

7. Verification that the specified transaction ID exists, points to your collateral address, is unspent and is equal to exactly 1000 Hilux. Messages in case of failure:  
  ![Could not find the specified transaction id](doc/img/startmn-tx-warning.png)  
  ![Collateral transaction output should equal 1000 Hilux](doc/img/startmn-collateral-warning.png)  
  If you decide to continue anyway, you probably won't be able to successfully start your masternode.

8. Verification at the Hilux network level that the specified transaction ID is valid. Message in case of failure:  
  ![Masternode broadcast message decode failed](doc/img/startmn-incorrect-tx-error.png)

9. After completing all pre-verification, the application will ask you whether you want to continue:  
  ![Press OK to broadcast transaction](doc/img/startmn-broadcast-query.png)  
  This is the last chance to stop the process.

10. Sending the `start masternode` message. Success returns the following message:  
  ![Successfully relayed broadcast message](doc/img/startmn-success.png)  
  In case of failure, the message text may vary, depending on the nature of the problem. Example:  
  ![Failed to start masternode](doc/img/startmn-failed-error.png)

## Transferring masternode earnings

DMT version 0.9.4 and above allows you to transfer your masternode earnings. Unlike other Hilux wallets, DMT gives you a 100% control over which *unspent transaction outputs* (utxo) you want to transfer. This has the same effect as the `Coin control` functionality implemented in the *Hilux Core* wallet.

The `Transfer funds` window shows all *UTXOs* of the currently selected Masternode (mode 1), all Masternodes in current configuration (mode 2) or any address controlled by a hardware wallet (mode 3). All *UTXOs* not used as collateral are initially selected. All collateral *UTXOs* (1000 Hilux) are initially hidden to avoid unintentionally spending collateral funds and thus breaking MN. You can show these hidden entries by unchecking the `Hide collateral utxos` option.

To show the `Transfer funds` window, click the `Tools` menu. Then, from the popup menu choose:
 * `Transfer funds from current Masternode's address` (mode 1)
 * `Transfer funds from all Masternodes addresses` (mode 2)
 * `Transfer funds from any address` (mode 3)

The same you can achieve by clicking of the three buttons from the right side of the app's toolbar:
![Transfer masternode funds window](doc/img/dmt-transfer-funds-tool-buttons.png)

Transferring funds from masternode collateral addresses (mode 2):  
![Transfer masternode funds window](doc/img/dmt-transfer-funds.png)

Transferring funds from any address controlled by a hardware wallet, using BIP32 path as an input (mode 3):  
![Transfer funds from any address window](doc/img/dmt-transfer-funds-any-address.png)

and using *wallet account* as an input (mode 3):  
![Transfer funds from any address window](doc/img/dmt-transfer-funds-any-address-account.png)

> Important: rows with a red font in the *Confirmations* column and a gray background are related to so-called *coinbase* transactions, that don't have the required number of confirmations to forward them. You should restrain from sending them and wait for them to receive at least 100 confirmations. 

To send funds, select all *UTXOs* you wish to include in your transaction, enter the details of the recipient(s), verify the transaction fee and click the `Prepare Transaction` button on the bottom: 
![Broadcast signed transaction confirmation](doc/img/dmt-transfer-funds-select-utxos.png)

After signing the transaction with your hardware wallet, the application will display a summary and will ask you for confirmation for broadcasting the signed transaction to the Hilux network.  
![Broadcast signed transaction confirmation](doc/img/dmt-transfer-funds-broadcast.png)

After clicking `Send Transaction`, the application broadcasts the transaction and then shows a confirmation with a transaction ID as a hyperlink directing to a Hilux block explorer:  
![Transaction sent](doc/img/dmt-transfer-funds-confirmation.png)

## Signing messages with a hardware wallet

To sign a message with your hardware wallet, click the `Tools` button and then select the `Sign message with HW for current Masternode's address` menu item. The `Sign message` window appears:  
![Sign message window](doc/img/dmt-hw-sign-message.png)

## Changing hardware wallet PIN/passphrase

Click the `Tools` button and select the `Hardware wallet PIN/Passphrase configuration` item. The following window will appear to guide you through the steps of changing the PIN/passphrase:  
![Hardware wallet setup window](doc/img/dmt-hardware-wallet-config.png)

## Downloads

This application is written in Python 3, but requires several additional libraries to run. These libraries in turn require the installation of a C++ compiler. All in all, compiling DMT from source is not trivial for non-technical people, especially the steps carried out under Linux (though this will be documented soon).

For this reason, in addition to providing the source code on GitHub, binary versions for all three major operating systems - macOS, Windows (32 and 64-bit) and Linux - are available for download directly. The binaries are compiled and tested under the following OS distributions:
* Windows 7 64-bit
* macOS 10.13.2 High Sierra
* Linux Debian Jessie

Binary versions of the latest release can be downloaded from: https://github.com/Bertrand256/hilux-masternode-tool/releases/latest.

### Verification of the application binary files
Beginning with version 0.9.15, each binary file forming part of a release has a corresponding signature file that you can use to verify the authenticity of the downloaded binary file (to ensure it has not been corrupted or replaced with a counterfeit) and confirm that it has been signed by the application author (Keybase user: bertrand256).

The verification method described below is based on use of the Keybase application, so if you have not already done so, download the installer from https://keybase.io/download and install the app.

**Verification steps**

1. Open your OS command line terminal

2. Change the current directory to the folder where the DMT release files were downloaded:

   `cd /Users/<username>/dmt-dist`

   After invoking the *list directory* command (`ls` for Mac/Linux, `dir` for Windows ) you should see both the archived executable (.tar.gz, .zip) and the corresponding signature file (.asc):

   ```
   <username>: ls
   HiluxMasternodeTool_0.9.15.mac.zip    HiluxMasternodeTool_0.9.15.mac.zip.asc
   ```


3. Verify the signature by executing the following command:

   ```
   keybase pgp verify -d HiluxMasternodeTool_0.9.15.mac.zip.asc -i HiluxMasternodeTool_0.9.15.mac.zip -S bertrand256
   ```

   You should see something similar to the following if verification was successful:

   ```
   ▶ INFO Identifying bertrand256
   ✔ public key fingerprint: 8044 DCB5 D241 349F 815F 9522 0D8F 46CA A62F 7EFD
   ฿  bitcoin 1BkMySjL6K72LWX82SVVmgCuZCuZ3DfuxS
   ฿  zcash t1byMTzteuJ8n1XUwxBzYXXV8qVv5kh3bFn
   ✔ "bertrand256" on reddit: https://www.reddit.com/r/KeybaseProofs/comments/688799/my_keybase_proof_redditbertrand256/ [cached 2018-01-19 10:27:37 CET]
   ✔ "bertrand256" on github: https://gist.github.com/781be961e33042c68d617782b0060844 [cached 2018-01-19 10:27:37 CET]
   Signature verified. Signed by bertrand256 1 hour ago (2018-01-19 11:46:05 +0100 CET).
   PGP Fingerprint: 8044dcb5d241349f815f95220d8f46caa62f7efd.
   ```


changelog.md
## [0.9.30] - 2021-10-17
**Changed**
- Proposals: fixing user errors regarding the "URL" field for 
addresses referring to HiluxCentral, a recent common mistake made by proposal owners.

## [0.9.29] - 2021-09-24
**Fixed**
- Fixed the "name 'fg_color' is not defined" error in the wallet tx confirmation dialog.
- Improving visibility of hyperlinks in dark mode (will be continued in several other dialogs).
- Updating dockerfile (Ubuntu based) for building the app.

## [0.9.28] - 2021-09-14
**Added**
- UI dark mode option independent of the OS settings (*Config->Miscellaneous->Use UI dark mode*).  
- New toolbar icons with enhanced visual consistency (kudos to **Doeke Koedijk** for creating them).
- Possibility to use keypad when entering PIN for Trezor devices. 

**Fixed**
- Support for macOS dark theme.
- Fixed an issue of the "Masternode status details" option disappearing after its usage.
- Fixed several other, minor visual issues.

## [0.9.27] - 2021-05-18
###Version with screenshots: [see here](doc/v0.9.27-changes.md)
**Added**
- A new way of presenting masternode data in the main window - masternode list.  
- Masternode details moved to a separate page/view.
- A view showing a set of information about the Hilux network.  
- Retrieving masternode information from the Hilux network based on IP/port when adding 
  a new masternode entry.
- A new window acting as a container for all application tools (Toolbox).  
- Hardware wallet settings 
  - Interface for the "Passphrase always on device" option for Trezor T.
  - Interface for setting the "Wipe code" option for Trezor One/T - an additional PIN, the 
    entering of which will wipe the device memory ([details](https://wiki.trezor.io/User_manual:Wipe_code)).
  - Interface for configuring an additional method of securing Trezor T devices - an SD 
    card that must be present to start the device ([details](https://wiki.trezor.io/User_manual:SD_card_protection)).
- Interface to the wallet seed recovery method for Trezor One with increased security, based on the word matrix.    
- Tool to help creating the "rpcauth" string, which allows starting RPC interface on a server without leaving 
  password to the interface on the server.
- Removing the strict setting of the hardware wallet type in the app configuration in favor of the dynamic selection 
  of the device from the currently connected ones.
- Adapting to the latest versions of communication libraries for Trezor, Keepkey and Ledger and 
  for PyQt.
- Adapting communication module to Hilux v0.17.
- Other, numerous improvements and fixes.

## [0.9.26-hotfix4] - 2021-02-18
**Fixed**
- Fixed an error occurring when signing transactions with Trezor using the latest firmware.
- Fixed an error in the firmware update wizard for Trezor devices.

## [0.9.26-hotfix2] - 2019-09-09

**Fixed**
- Fixed an issue with saving the ssh authentication method.

## [0.9.26-hotfix1] - 2019-08-20

**Fixed**
- Proposals: remove votes from the cache that no longer exist on the network.

## [0.9.26] - 2019-08-19
**Added**
- Commandline parameters for changing the sig_time random offset range.
- Ability to explicitly specify the authentication methods for an SSH tunnel from: username/password, RSA private key 
pair and ssh agent.
- Workaround for Trezor connection issues (LIBUSB) after Windows update #1903.
- Showing a message if duplicate masternode information in the configuration may prevent voting. 

**Fixed**
- Issue "'WalletDlg' object has no attribute 'config'" when signing message from the wallet dialog.

## [0.9.25-hotfix2] - 2019-07-15

**Fixed**
- Issue "'WalletDlg' object has no attribute 'config'" when signing message from the wallet dialog.

## [0.9.25-hotfix1] - 2019-07-12

**Fixed**
- Issue with calculating the next payment block when a masternode received PoSeBan in the past and its PoSeRevivedBlock 
is less than the last payment block.

## [0.9.25] - 2019-07-01
**Added**
- Support for the KeepKey v6.x firmware.
- Possibility to limit the Trezor transport methods using commandline parameters.
- Signing messages with owner/voting key.
- Export/import of the configuration.

**Fixed**
- A bug related to access to uninitialized variable.

## [0.9.24] - 2019-05-24

**Added**
- Support for the "update service" feature (mn IP/port, operator payout address).
- Support for the "revoke operator" feature.
- Additional encryption (RSA) of data sent over the Internet for protx RPC calls.

**Changed**
- Additional information in the status area, mainly concerning diagnosis of problems with masternode.


## [0.9.23-hotfix3] - 2019-05-08

**Fixed**
- Malfunctioning/incorrect "details" link in the "Update payout address" dialog (the manual commands area).
- Issue occurring during sending a transaction after switching hw identities (an empty message box).

**Changed**
- Extended information about transaction recipients in the transaction window.

## [0.9.23-hotfix2] - 2019-05-03

**Fixed**
- Fix of the "invalid operator public key" error occurring in the
"Update operator key" window when using the public key option (Kudos to 
@Thiagokroger from Node40 for precise hints).

## [0.9.23-hotfix1] - 2019-05-02

**Fixed**
- Issue with "Locate collateral" feature (an empty error message).
- Switching the active RPC connection to another after encountering "Unknown error" and "401 Unauthorized" 
errors.

## [0.9.23] - 2019-04-27

**Added**
- The feature of restoring configuration from backup.
- GUI for update_registrar (changing payout address, operator/voting key).
- Extending the status area with additional information, including
the estimated date of the next payment.

**Changed**
- Removed the pre-spork 15 code.
- Fixes in the proposals dialog: support for v0.14-beta, issues with number of payment cycles. 
- Clearing the pre-spork 15 voting results from the app cache to suggest 
users the need to re-submit their votes.
- Fixed some stability issues in the wallet dialog.

## [0.9.22] - 2019-02-24

**Added**

src/app_defs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Author: Bertrand256
# Created on: 2018-03
import collections
import logging
from enum import Enum
from typing import List

from PyQt5.QtGui import QColor

APP_NAME_SHORT = 'HiluxMasternodeTool'
APP_NAME_LONG = 'Hilux Masternode Tool'
APP_DATA_DIR_NAME = '.dmt'
PROJECT_URL = 'https://github.com/Bertrand256/hilux-masternode-tool'
FEE_DUFF_PER_BYTE = 1
MIN_TX_FEE = 1000
SCREENSHOT_MODE = False
DEBUG_MODE = False
DEFAULT_LOG_FORMAT = '%(asctime)s %(levelname)s|%(name)s|%(threadName)s|%(filename)s|%(funcName)s|%(message)s'
KnownLoggerType = collections.namedtuple('KnownLoggerType', 'name external')
APP_PATH = ''
APP_IMAGE_DIR = ''
BROWSER_USER_AGENT = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) ' \
                     'Chrome/88.0.4324.152 Safari/537.36'

COLOR_WARNING_STR = '#e65c00'
COLOR_WARNING = QColor(COLOR_WARNING_STR)
COLOR_ERROR_STR = 'red'
COLOR_ERROR = QColor(COLOR_ERROR_STR)


class AppTextMessageType(Enum):
    INFO = 'info'
    WARN = 'warn'
    ERROR = 'error'


class DispMessage(object):
    NEW_VERSION = 1
    DASH_NET_CONNECTION = 2
    OTHER_1 = 3
    OTHER_2 = 4

    def __init__(self, message: str, type: AppTextMessageType):
        """
        :param type: 'warn'|'error'|'info'
        :param message: a message
        """
        self.message = message
        self.type: AppTextMessageType = type
        self.hidden = False


def get_note_url(note_symbol):
    """
    Returns an URL to a project documentation page related to the note symbol passed as an argument.
    :param note_symbol: Symbol of the note, for example: DMT00001
    :return: URL
    """
    return PROJECT_URL + f'/blob/master/doc/notes.md#note-{note_symbol.lower()}'


def get_doc_url(doc_file_name):
    """
    Returns an URL to a project documentation page.
    :return: URL
    """
    return PROJECT_URL + f'/blob/master/doc/{doc_file_name}'


__KNOWN_LOGGERS = [
    KnownLoggerType(name='dmt.wallet_dlg', external=False),
    KnownLoggerType(name='dmt.bip44_wallet', external=False),
    KnownLoggerType(name='dmt.hiluxd_intf', external=False),
    KnownLoggerType(name='dmt.db_intf', external=False),
    KnownLoggerType(name='dmt.proposals', external=False),
    KnownLoggerType(name='dmt.ext_item_model', external=False),
    KnownLoggerType(name='dmt.hw_intf', external=False),
    KnownLoggerType(name='dmt.hw_intf_trezor', external=False),
    KnownLoggerType(name='dmt.reg_masternode', external=False),
    KnownLoggerType(name='dmt.transaction_dlg', external=False),
    KnownLoggerType(name='dmt.app_cache', external=False),
    KnownLoggerType(name='BitcoinRPC', external=True),
    KnownLoggerType(name='urllib3.connectionpool', external=True),
    KnownLoggerType(name='trezorlib.transport', external=True),
    KnownLoggerType(name='trezorlib.transport.bridge', external=True),
    KnownLoggerType(name='trezorlib.client', external=True),
    KnownLoggerType(name='trezorlib.protocol_v1', external=True),
]


def get_known_loggers() -> List[KnownLoggerType]:
    ll = []
    # add existing loggers which are not known: some new libraries (or new versions) can introduce new
    # loggers
    for lname in logging.Logger.manager.loggerDict:
        l = logging.Logger.manager.loggerDict[lname]
        if isinstance(l, logging.Logger):
            if not any(lname in name for name in __KNOWN_LOGGERS):
                __KNOWN_LOGGERS.append(KnownLoggerType(name=lname, external=True))
    return __KNOWN_LOGGERS[:]